### DISCUSSION TITLE
PDF player accessibility

### DISCUSSION THREAD

**Discussion by HarishGangula** [Sunbird-Knowlg/Community#23] [General Discussions]
**Problem Statement:**

The current implementation of the Sunbird PDF player is not completely accessible compliant. For eg., although the screen reader reads the complete PDF document it isn’t able to understand the structure of the document (H1, H2 labels) and the punctuation. Hence making it difficult for users with visual impairment to understand the content in the intended way.  Currently Player uses the default render of PDF JS which renders all text as span HTML tags and other media elements using canvas HTML tag without any attributes, This makes it difficult for screen readers to present in the intended way. 

**More details about Sunbird PDF player:**

Sunbird PDF player is built using the PDF JS open-source library. This open-source library doesn’t support these capabilities. https://bugzilla.mozilla.org/show_bug.cgi?id=861157 is the open issue for the same

**What others have done?:**

Linkedin also uses the PDF JS open-source library to render PDF documents. To address this problem, they have built a custom viewer (which is not yet open-sourced). More details about how they have done it can be found here: https://engineering.linkedin.com/blog/2019/04/under-the-hood--learning-with-documents
This capability was effort-intensive - took around a year for them to implement it.
Note: we couldn’t find any additional information about how other adopters of PDF JS open-source library have addressed this.

We have gone through the adobe PDF standards (https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/pdfreference1.0.pdf) and from our analysis it looks like to enable this capability we need to: 

1. Understand the pdf specification for creating the renderer using the pdf standard document 
2. After the understanding it, we have to build the pdf renderer using the PDF JS library

We are seeking help from the community to guide us if there are any other options we need to consider to address this problem or any other reference open-source players built using the PDF JS library.

@pramodkvarma   @vrayulu @SanthoshVasabhaktula @Krishnaj20 @vinukumar-vs @ChakshuGautam @madhucr @pallakartheekreddy
@sharathkashyap @rhwarrier @krgauraw @AmiableAnil @Jayaprakash8887  @coolbung @luckynvdu

**Comment by pramodkvarma** [comment]
I am surprised that there is no open source PDF viewer for browser!! How do browsers render? Could we leverage the native rendering capability of browser itself instead of looking for a programmable PDF viewer library? Just wondering.

**Reply by HarishGangula** [reply]
Yes, @pramodkvarma  there are no open-source PDF viewers which support accessibility (https://en.wikipedia.org/wiki/Computer_accessibility) inbuilt. PDF JS is the most popular open source PDF viewer and it does not have support for accessibility.

Browsers have native support for PDF which they use to render PDF via an <embed> tag or iframe. Chrome's built-in PDF viewer is itself a plugin, running as native code via the [Pepper Plugin](http://en.wikipedia.org/wiki/Google_Native_Client#Pepper) API (PPAPI), Firefox includes PDF.js as its PDF viewer, and Safari has built-in PDF support. But they also do not provide accessibility features (to help disabled users).

So, we need to build a layer on top of PDF JS or browser native capability to enable accessibility features. However, it looks very complex to implement based on what LinkedIn has done to enable accessibility within PDFs.

**Comment by pramodkvarma** [comment]
Oh, that's unfortunate. Thanks for clarifying @HarishGangula. Let us see if someone from community can put up their hand to help enhance PDF JS.

**Comment by surendrasinghs** [comment]
This is also a very important "white space" that we have discovered. It would be of great value to make this available as open-source to help further the inclusivity agenda. I am pinning this thread to the discussion forum to highlight this to the community.

**Reply by pramodkvarma** [reply]
Indeed! Maybe take it to LinkedIn and twitter as well and point them to this thread. Let us see if we can find someone.
